src/storage/content_directory.rs

```diff
@@ -0,0 +1,145 @@
+
+pub trait Trait: timestamp::Trait + system::Trait + DOTRTrait + MaybeDebug {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    type MetadataId: Parameter + Member + SimpleArithmetic + Codec + Default + Copy
+        + As<usize> + As<u64> + MaybeSerializeDebug + PartialEq;
+
+    // Schema ID should be defined in a different Trait in future
+    type SchemaId: Parameter + Member + SimpleArithmetic + Codec + Default + Copy
+        + As<usize> + As<u64> + MaybeSerializeDebug + PartialEq;
+
+
+    type IsActiveMember: IsActiveMember<Self>;
+}
+
+static MSG_CREATOR_MUST_BE_MEMBER: &str = "Only active members may create content!";
+static MSG_INVALID_SCHEMA_ID: &str = "The metadata schema is not known or invalid!";
+static MSG_METADATA_NOT_FOUND: &str = "Metadata with the given ID cannot be found!";
+static MSG_ONLY_OWNER_MAY_PUBLISH: &str = "Only metadata owners may publish their metadata!";
+
+const DEFAULT_FIRST_METADATA_ID: u64 = 1;
+
+#[derive(Clone, Encode, Decode, PartialEq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum MetadataState
+{
+    Draft,
+    Published,
+}
+
+impl Default for MetadataState {
+    fn default() -> Self {
+        MetadataState::Draft
+    }
+}
+
+// Content metadata contains two fields: one is the numeric ID of the metadata
+// scheme to use, which is currently related to off-chain definitions ONLY.
+// The other is the serialized metadata.
+#[derive(Clone, Encode, Decode, PartialEq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Metadata<T: Trait> {
+    pub schema: T::SchemaId,
+    pub metadata: Vec<u8>,
+    pub origin: T::AccountId,
+    pub state: MetadataState,
+}
+
+decl_event! {
+    pub enum Event<T> where
+        <T as Trait>::MetadataId
+    {
+        MetadataDraftCreated(MetadataId),
+        MetadataPublished(MetadataId),
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as ContentDirectory {
+        // Start at this value
+        pub FirstMetadataId get(first_metadata_id) config(first_metadata_id): T::MetadataId = T::MetadataId::sa(DEFAULT_FIRST_METADATA_ID);
+
+        // Increment
+        pub NextMetadataId get(next_metadata_id) build(|config: &GenesisConfig<T>| config.first_metadata_id): T::MetadataId = T::MetadataId::sa(DEFAULT_FIRST_METADATA_ID);
+
+        // Mapping of Data object types
+        pub MetadataMap get(metadata): map T::MetadataId => Option<Metadata<T>>;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event<T>() = default;
+
+        pub fn add_metadata(origin, schema: T::SchemaId, metadata: Vec<u8>)
+        {
+            // Origin has to be a member
+            let who = ensure_signed(origin).clone().unwrap();
+            if !T::IsActiveMember::is_active_member(&who) {
+                return Err(MSG_CREATOR_MUST_BE_MEMBER);
+            }
+
+            // TODO in future, we want the schema IDs to correspond to a schema
```

Comment:
Makes sense to mention a corresponding issue here in this TODO-comment:
https://github.com/Joystream/substrate-runtime-joystream/issues/20